src/nbl/video/CVulkanConnection.h

```diff
@@ -42,79 +59,86 @@ class CVulkanConnection final : public IAPIConnection
-        // volkLoadInstanceOnly(m_instance);
-        volkLoadInstance(m_instance);
-
-        vkCreateDebugUtilsMessengerEXT(m_instance, &debugUtilsMessengerCreateInfo, nullptr, &m_debugMessenger);
-        assert(m_debugMessenger);
-        
-        uint32_t devCount = 0u;
-        vkEnumeratePhysicalDevices(m_instance, &devCount, nullptr);
-        core::vector<VkPhysicalDevice> vkphds(devCount, VK_NULL_HANDLE);
-        vkEnumeratePhysicalDevices(m_instance, &devCount, vkphds.data());
-
-        m_physDevices = core::make_refctd_dynamic_array<physical_devs_array_t>(devCount);
-        for (uint32_t i = 0u; i < devCount; ++i)
+
+        physical_devs_array_t physicalDevices = core::make_refctd_dynamic_array<physical_devs_array_t>(physicalDeviceCount);
+        for (uint32_t i = 0u; i < physicalDeviceCount; ++i)
         {
-            (*m_physDevices)[i] = core::make_smart_refctd_ptr<CVulkanPhysicalDevice>(vkphds[i], core::smart_refctd_ptr(m_system), std::move(m_GLSLCompiler));
+            (*physicalDevices)[i] = core::make_smart_refctd_ptr<CVulkanPhysicalDevice>(
+                vk_physicalDevices[i], core::smart_refctd_ptr(sys),
+                core::make_smart_refctd_ptr<asset::IGLSLCompiler>(sys.get()));
         }
+
+        return core::make_smart_refctd_ptr<CVulkanConnection>(vk_instance, physicalDevices, vk_debugMessenger);
     }
 
-    E_API_TYPE getAPIType() const override { return EAT_VULKAN; }
+    VkInstance getInternalObject() const { return m_vkInstance; }
 
     core::SRange<const core::smart_refctd_ptr<IPhysicalDevice>> getPhysicalDevices() const override
     {
-        return core::SRange<const core::smart_refctd_ptr<IPhysicalDevice>>{ m_physDevices->begin(), m_physDevices->end() };
+        return core::SRange<const core::smart_refctd_ptr<IPhysicalDevice>>{ m_physicalDevices->begin(), m_physicalDevices->end() };
     }
 
-    core::smart_refctd_ptr<ISurface> createSurface(ui::IWindow* window) const override;
+    E_API_TYPE getAPIType() const override { return EAT_VULKAN; }
 
-    VkInstance getInternalObject() const { return m_instance; }
+    // Todo(achal)
+    IDebugCallback* getDebugCallback() const override { return nullptr; }
+
+// Todo(achal): Remove
+// private:
+
+    using physical_devs_array_t = core::smart_refctd_dynamic_array<core::smart_refctd_ptr<IPhysicalDevice>>;
+
+    CVulkanConnection(VkInstance instance, physical_devs_array_t physicalDevices,
+        VkDebugUtilsMessengerEXT debugUtilsMessenger = VK_NULL_HANDLE)
+        : m_vkInstance(instance), m_physicalDevices(physicalDevices),
+        m_vkDebugUtilsMessengerEXT(debugUtilsMessenger)
+    {}
 
-protected:
     ~CVulkanConnection()
     {
-        vkDestroyDebugUtilsMessengerEXT(m_instance, m_debugMessenger, nullptr);
-        vkDestroyInstance(m_instance, nullptr);
+        vkDestroyDebugUtilsMessengerEXT(m_vkInstance, m_vkDebugUtilsMessengerEXT, nullptr);
+        vkDestroyInstance(m_vkInstance, nullptr);
```

Comment:
constructor and destructor need to be `protected`

Reply:
Resolved